### ISSUE TITLE
Trial Design/Conduct Interface Development

### ISSUE THREAD

**Issue by xlai** [xlai/dosefinderhub#4]
- [x] Integration of questionnaire responses into main UI
- [ ] Design UI for trial design/conduct display
- [x] Implement comparison feature using simulation studies

**Comment by xlai** [comment]
@eduardo-vie : please coordinate with both @sdennett20 @jakeholton to see what variables you would need and you can provide. Save the list in a plain text file under the doc for now

**Comment by eduardo-vie** [comment]
Hi Xiaoran, thanks a lot for organizing the project and providing instructions.
I have some broad questions concerning project structure and task allocation.

- The way we were thinking about it when coming up with the Questionnaire yesterday, it would include general questions about the clinician's goals/constraints, such as "Would you like to have a flexible  TTL?" (no specific requests, e.g. to specify prior information).
- Sian would then find a way to utilize the answers to generate something like a trial design ranking, 1 being most recommended (e.g., CRM, followed by SPM, followed by CCD…). This ranking may or may not be displayed and explained to the user at the end of the Questionnaire UI.
- As such, the input from the Questionnaire UI to the Main UI would be this ranking, potentially as simple as a vector. In this case, it wouldn't really make sense for us to specify a list of input variables as suggested here.
- Based on the ranking, the main UI would ask the user for specific information to subsequently simulate each design (e.g., if the 1st ranked design is CRM, it will have to ask the user to specify prior info in one way or another).
- With these parameters collected, all ranked designs' simulations would run, and their operating characteristics would be explicitly compared.

Is this far from what you had in mind?
If not, my next step would be to start planning the Main UI and trial comparison feature with Jake.
Thanks again!

**Comment by eduardo-vie** [comment]
Notes on Main UI, as per 17/8 catch up:

- User will start in a Configurations tab. There, they will see 3 drop down columns, one for each simulated design. Once they click to drop down a column, the column will be sectioned based on info type (clinical parameters, simulation parameters etc.). These sections will include the same design-specific questions as the latter portion of the Questionnaire UI. If the user has completed the Questionnaire UI, the previous answers will be fed into and displayed in this tab. If they have not, they will use this tab to provide this information.
- The user will then be able to flick through 3 other tabs, one for each simulated design, containing simulation outputs (DTPs, average no participants per dose, probabilities of dose selection etc.). They should also include some sort of summarized cross-design comparison output.
- The 5th tab will be a comparison tab, with more detailed cross-design comparison outputs.

**Comment by eduardo-vie** [comment]
I have created the the backbone of the backbone of the Main UI (Trial Design/Conduct UI), including its 5 tabs. I have also made a functional mock First Tab (Configurations Tab).
Depending on the ranking vector I get from @sdennett20, such as c("3plus3", "crm", "other"), the Configurations tab will display 3 accordingly ordered columns showing and allowing the user to update their previously specified design-specific parameters. Changing the recommended designs (which we can't do yet because we only have these three) or their order (which I am already doing) changes column content and order.
I made it so that each design's specific parameters can be defined/added in an individual function, which is selected based on the ranking, and then called within the larger UI function. This should facilitate expansion and customization as we advance in the project.
Here's the rough code and a screenshot of the mock Configurations Tab.
My next step will be making mock versions of the other tabs, though this should be harder until we specify what simulation results/metrics we are calculating and using for comparison.

[Main_UI_V2_text.txt](https://github.com/xlai/dosefinderhub/files/12397094/Main_UI_V2_text.txt)

![image](https://github.com/xlai/dosefinderhub/assets/141162230/de1354f1-b10d-43fb-b485-f1402953f99a)

**Comment by xlai** [comment]
Well done for coding up your first shiny app! The interface is in a good position for our first round of review. I have a few comments on the "parameter configuration" tab:

1. How will you be displaying the trial-specific and method-specific parameters that will be passed from Sian's questionnaire interface? One possibility is to have a tick box somewhere that hides/displays options that were pre-specified.
3. There should be a button that allows adding of additional scenarios, and maybe options to upload csv files for batch scenario simulation.
4. I also noticed that the parameters so far are mostly related to the running of simulations. If we are planning to run the same scenario for all methods, perhaps they are only needed to be specified once. 

The next step would be to add in the trial conduct interface, meaning the sequential assignment of dose levels to (synthetic) patients. We can discuss this tomorrow.

**Comment by eduardo-vie** [comment]
Thanks for the swift feedback!
1. I will display them in the columns, but I wanna make them drop-down as we discussed.
2. I was thinking of doing that in an easier but worse way (More scenarios? YES vs NO -> YES triggers more questions) or harder but better way (an empty table, each row being a scenario and each column being a dose, where user needs to fill out at least one row). I need to learn a way to code the latter but I much prefer it.
3. Yes, true. They are defined as deisng-specific in the variables spreadsheet though, mostly because we were thinking users may want to, e.g., run more simulations for one design than another. But we can def re-visit that.
See you tomorrow! :D

**Comment by xlai** [comment]
>  I need to learn a way to code the latter (which will probably have to be reactive?) but I much prefer it. And thanks for the suggestion to add a csv upload option.

```R
library(shiny)

ui <- fluidPage(
  titlePanel("Simulation Scenarios"),
  sidebarLayout(
    sidebarPanel(
      numericInput("num_rows", "Number of Rows:", value = 5),
      actionButton("add_rows", "Add Rows")
    ),
    mainPanel(
      tableOutput("scenario_table")
    )
  )
)

server <- function(input, output, session) {
  # Initialize the data frame
  initial_data <- data.frame(Dose = numeric(0), Scenario = character(0))
  
  # Reactive value to store the data
  data <- reactiveVal(initial_data)
  
  observeEvent(input$add_rows, {
    current_data <- data()
    new_rows <- data.frame(Dose = rep(0, input$num_rows), Scenario = rep("", input$num_rows))
    new_data <- rbind(current_data, new_rows)
    data(new_data)
  })
  
  output$scenario_table <- renderTable({
    data()
  })
}

shinyApp(ui, server)

```

A simple working example to get you started on reactive table.

**Comment by eduardo-vie** [comment]
Closing this issue because the following has been accomplished:
- The dynamic backbone of the **Main UI** has been coded. This includes: a **Configurations tab** with simulation parameter inputs (namely, a reactive, editable table for the user to enter each simulations scenario's "true" DLT rates) + dynamically-triggered, drop-down design-specific input columns; multiple **design/simulations tabs** (still blank); a **Conduct tab** (with some initial conduct info inputs)
- The Main UI has been made robust/dynamic for ranking vectors of any length and order
- The code has been somewhat modularized and generalized to facilitate future changes

Opening new issue (Trial Design/Conduct Interface Development - Second Iteration) to address the next steps, summarized as:
- Set up use of Sian's data frames and renderUI() function to remove mock variables and UI input specs from my code, streamlining it
- Add fixed non-design-specific column to Configurations tab
- Based on Jake's simulation outputs, plan and execute the single Design tab (simulation outputs tab). This should requiring structuring the relaying of my data frames to Jake
- Continue planning and developing the Conduct tab
- In summary, have a fully functional main UI featuring at least some inputs
